cuda_core/cuda/core/experimental/_memory.pyx

```diff
@@ -638,35 +622,67 @@ class DeviceMemoryResource(MemoryResource):
             self._ipc_handle_type = properties.handleTypes
             self._mempool_owned = True
             self._is_imported = False
+            self._uuid = None
+            self._alloc_handle = None
 
             err, self._mempool_handle = driver.cuMemPoolCreate(properties)
             raise_if_driver_error(err)
 
+            if opts.ipc_enabled:
+                self.get_allocation_handle() # enables Buffer.export, sets uuid
+
     def __del__(self):
         self.close()
 
     def close(self):
         """Close the device memory resource and destroy the associated memory pool if owned."""
-        if self._mempool_handle is not None and self._mempool_owned:
-            err, = driver.cuMemPoolDestroy(self._mempool_handle)
-            raise_if_driver_error(err)
+        if self._mempool_handle is not None:
+            try:
+                if self._mempool_owned:
+                    err, = driver.cuMemPoolDestroy(self._mempool_handle)
+                    raise_if_driver_error(err)
+            finally:
+                self._dev_id = None
+                self._mempool_handle = None
+                self._attributes = None
+                self._ipc_handle_type = _NOIPC_HANDLE_TYPE
+                self._mempool_owned = False
+                self._is_imported = False
+                self._uuid = None
+                self._alloc_handle = None
 
-            self._dev_id = None
-            self._mempool_handle = None
-            self._attributes = None
-            self._ipc_handle_type = _NOIPC_HANDLE_TYPE
-            self._mempool_owned = False
-            self._is_imported = False
 
-    @classmethod
-    def from_shared_channel(cls, device_id: int | Device, channel: IPCChannel) -> DeviceMemoryResource:
-        """Create a device memory resource from a memory pool shared over an IPC channel."""
-        device_id = getattr(device_id, 'device_id', device_id)
-        alloc_handle = channel._proxy._receive_allocation_handle()
-        return cls._from_allocation_handle(device_id, alloc_handle)
+    def __reduce__(self):
+        # If spawning a new process, serialize the resources; otherwise, just
+        # send the UUID, using the registry on the receiving end.
+        is_spawning = multiprocessing.context.get_spawning_popen() is not None
+        if is_spawning:
+            from ._device import Device
+            device = Device(self.device_id)
+            alloc_handle = self.get_allocation_handle()
```

Comment:
Isn't `get_allocation_handle` already called in the constructor? Can we just use `self._alloc_handle` here?